framework.cpp

```diff
@@ -0,0 +1,403 @@
+#ifndef _METRIC_FRAMEWORK_CPP
+#define _METRIC_FRAMEWORK_CPP
+
+#include <cmath>
+#include <vector>
+//#include <map>
+
+#include "3rdparty/blaze/Blaze.h"
+
+
+namespace metric {
+
+
+
+
+// --------
+// common type trait metafunctions
+
+
+
+template <typename>
+struct determine_container_type  // checks whether container is STL container (1) or Blaze vector (2)
+{
+    constexpr static int code = 0;
+};
+
+template <template <typename, typename> class Container, typename ValueType, typename Allocator>
+struct determine_container_type<Container<ValueType, Allocator>>
+{
+    constexpr static int code = 1; // STL
+};
+
+template <template <typename, bool> class Container, typename ValueType, bool F>
+struct determine_container_type<Container<ValueType, F>>
+{
+    constexpr static int code = 2; // blaze::DynamicVector
+};
+
+template <template <typename, int, int, int, int, int> class Container, typename ValueType, int _Rows, int _Cols, int _Options, int _MaxRows, int _MaxCols>
+struct determine_container_type<Container<ValueType, _Rows, _Cols, _Options, _MaxRows, _MaxCols>>
+{
+    constexpr static int code = 3; // Eigen::Array
+};
+
+
+template<typename C, int = determine_container_type<C>::code>
+struct determine_val_t  // determines type of element both for STL containers and Blaze vectors
+{
+    using type = void;
+};
+
+template<typename C>
+struct determine_val_t<C, 1>
+{
+    using type = typename C::value_type;
+};
+
+template<typename C>
+struct determine_val_t<C, 2>
+{
+    using type = typename C::ElementType;
+};
+
+template<typename C>
+struct determine_val_t<C, 3>
+{
+    using type = typename C::Scalar;
+};
+
+
+
+// --------
+// base classes
+
+
+template <typename rec_t>
+class MetricBase { // all metrics to be derived from
+
+public:
+
+    using Rec_T = rec_t;
+    using Val_T = determine_val_t<Rec_T>;
+
+    MetricBase<rec_t>() {}
+
+    double operator () (const rec_t & a, const rec_t & b) { // to be overridden in derived metrics
+        return 0;
+    }
+
+private:
+
+};
+
+
+
+
+template <typename Metric = MetricBase<double>>
+class Algorithm {
+
+public:
+
+    //using Rec_T = rec_t;
+    //using Val_T = determine_val_t<Rec_T>;
+
+    Algorithm<Metric>(Metric metric_) : metric(metric_) {}
+
+    typename Metric::Rec_T process_single(typename Metric::Rec_T in) {
+        return in;
+    }
+
+    template <typename set_t>
+    set_t process_batch(set_t in) {
```

Comment:
it is not clear how this two methods (`process_single` and `process_batch`) plays later with concrete algorithms.
I assume different algorithms has different number of inputs (1, 2, 3) as there is no way to disable base class methods nicely this could result in nasty bugs.

Reply:
all mapping and transform algorithms encode single records, for example a vector<double> and return a double. here we can provide simple batch processing for n records.

Reply:
my point is that there is no use for default method in base algorithm class, as its implementation will always be wrong (no way to call single records process, unless we make it virtual), and because we still can accidentally call this default method I think it is rather error-prone.

and what about `correlation` module?

Reply:
We can have different base template for different algorithm group, what do you think?

Reply:
For the now it looks like there is no need in base algorithm classes class at all. Let's make each algorithm independent, and only keep some standard. We can create a base class for any group of algorithms at ant time, if we need to place any common code there.

Reply:
@maxfilippov I agree, we could put standard inside README in each module folder

Reply:
ok, once we finish establishing the standard, i'll try to describe it, and we will be able to work with this description and update it if needed